EngineLayer/Localization/LocalizationEngine.cs

```diff
@@ -26,36 +28,60 @@ public LocalizationEngine(IEnumerable<PeptideSpectralMatch> allResultingIdentifi
         protected override MetaMorpheusEngineResults RunSpecific()
         {
             // don't try to localize mass differences for ambiguous peptides
-            foreach (PeptideSpectralMatch psm in AllResultingIdentifications.Where(b => b.FullSequence != null))
-            {
-                // Stop loop if canceled
-                if (GlobalVariables.StopLoops) { break; }
-
-                MsDataScan scan = MyMsDataFile.GetOneBasedScan(psm.ScanNumber);
-                Ms2ScanWithSpecificMass scanWithSpecificMass = new Ms2ScanWithSpecificMass(scan, psm.ScanPrecursorMonoisotopicPeakMz, psm.ScanPrecursorCharge, psm.FullFilePath, CommonParameters);
-                PeptideWithSetModifications peptide = psm.BestMatchingPeptides.First().Peptide;
-                double massDifference = psm.ScanPrecursorMass - peptide.MonoisotopicMass;
-
-                // this section will iterate through all residues of the peptide and try to localize the mass-diff at each residue and report a score for each residue
-                var localizedScores = new List<double>();
-                for (int r = 0; r < peptide.Length; r++)
+            PeptideSpectralMatch[] unambiguousPsms = AllResultingIdentifications.Where(b => b.FullSequence != null).ToArray();
+
+            double psmsSearched = 0;
+            int oldPercentProgress = 0;
+
+            ReportProgress(new ProgressEventArgs(oldPercentProgress, "Localizing mass-differences... ", NestedIds));
+
+            Parallel.ForEach(Partitioner.Create(0, unambiguousPsms.Length),
+                new ParallelOptions { MaxDegreeOfParallelism = CommonParameters.MaxThreadsToUsePerFile },
+                (range, loopState) =>
                 {
-                    // create new PeptideWithSetMods with unidentified mass difference at the given residue
-                    PeptideWithSetModifications peptideWithLocalizedMassDiff = peptide.Localize(r, massDifference);
+                    for (int i = range.Item1; i < range.Item2; i++)
+                    {
+                        PeptideSpectralMatch psm = unambiguousPsms[i];
+
+                        // Stop loop if canceled
+                        if (GlobalVariables.StopLoops) { break; }
```

Comment:
this code is the same as before except parallelized, with search progress reported